### ISSUE TITLE
Regression: false positive from unreachable_pub lint

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#102352]
### Code

I tried this code:

```rust
#![deny(unreachable_pub)]

pub use crate::builder::Foo;

mod builder {
    pub struct Foo(pub(crate) ());
}
```

I expected to see no output from the compiler.

Instead, this happened:

```rust
djc-2021 main test-rs $ cargo +nightly c
    Checking test-rs v0.1.0 (/Users/djc/src/test-rs)
error: unreachable `pub` item
 --> src/lib.rs:3:9
  |
3 | pub use crate::builder::Foo;
  | ---     ^^^^^^^^^^^^^^^^^^^
  | |
  | help: consider restricting its visibility: `pub(crate)`
  |
```

## cargo-bisect-rustc ouptut

searched nightlies: from nightly-2022-09-22 to nightly-2022-09-27
regressed nightly: nightly-2022-09-25
searched commit range: https://github.com/rust-lang/rust/compare/4a14677239bbd9a3892d165ee24513c1a05820e4...3f83906b30798bf61513fa340524cebf6676f9db
regressed commit: https://github.com/rust-lang/rust/commit/6580010551063718462f9dfe41c9490d92994d0e

This is a roll-up, #102109 seems the most likely culprit. cc @petrochenkov @oli-obk 

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.4</summary>

Host triple: aarch64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start=2022-09-22 
```
</details>

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

**Comment by djc** [comment]
Looks like tokio is hitting this, too: https://github.com/tokio-rs/tokio/actions/runs/3138494294/jobs/5097884393.

**Comment by apiraino** [comment]
WG-prioritization assigning priority, probably a P-high since the lint malfunction is blocking valid code (iiuc)

@rustbot label -I-prioritize P-high T-compiler

**Comment by oli-obk** [comment]
@petrochenkov any ideas how to address this or should I start digging in to it

**Comment by petrochenkov** [comment]
In HIR `pub use crate::builder::Foo;` desugars into two items (import in type namespace and import in value namespace).
One of them is externally reachable and another is not (before https://github.com/rust-lang/rust/pull/102109 both of them were incorrectly marked as externally reachable).

Code in `compiler\rustc_lint\src\builtin.rs` should somehow account for this desugaring and not report the `pub` as unnecessary if at least one of the imports is reachable.

**Comment by oli-obk** [comment]
I have been unable to fix this or apply a revert. @petrochenkov can you look at implementing this before the beta cutoff?

**Comment by petrochenkov** [comment]
Reverted in https://github.com/rust-lang/rust/pull/103249.
